### PR TITLE
Implements a module for working with Trains

### PR DESCRIPTION
**Pull request by AndyBursh**
Hi Afforess and contributors,

As discussed in #44, here is the pull request for a trains module. There are a couple of things I think are worth noting:

1. I've followed your lead on module structure from the Event module. My implementation looks solid to me, but I'd appreciate any feedback on how I've implemented private members (I've only learned Lua in the last week and a bit, so I'm not familiar with its idioms yet)

2. I haven't really included any examples of how modders would use this module in conjunction with other modules (eg the Entity module) because I didn't really know where to put them. If you can provide guidance on that, I'll add them.

3. I did originally split the tests of the modules functions and its behaviour upon being loaded in to two files, but I found I had trouble getting both sets of tests to work smoothly and I was repeating myself a lot in order to get the initial state right. For those reasons, I merged them in to a single file.

I feel this is a good starting point for the trains module; I'm sure it'll need some tweaks and some additions as we learn how to use it in a real mod, but I'm happy with this base.

Regards,
Andy

### CODE REVIEW

File: stdlib/trains/trains.lua [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+
+-- When developers  load this module, we need to
+-- attach some new events
+
+-- Filters events related to entity_type
+-- @tparam string event_parameter The event parameter to look inside to find the entity type
+-- @tparam string entity_type The entity type to filter events for
+-- @tparam callable callback The callback to invoke if the filter passes. The object defined in the event parameter is passed.
+function filter_event(event_parameter, entity_type, callback)
```
Comment (Afforess) [inline stdlib/trains/trains.lua]: I'd make this a local function. Without the local modifier, this function is part of the global namespace.

File: stdlib/trains/trains.lua [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        train.locomotives and
+        (#train.locomotives.front_movers > 0 or #train.locomotives.back_movers > 0)
+    then
+        return train.locomotives.front_movers and train.locomotives.front_movers[1] or train.locomotives.back_movers[1]
+    end
+end
+
+-- Creates a registry of known trains
+-- @return table A mapping of train id to LuaTrain object
+function create_train_registry()
```
Comment (Afforess) [inline stdlib/trains/trains.lua]: Same as https://github.com/Afforess/Factorio-Stdlib/pull/46/files#r78116046

File: stdlib/trains/trains.lua [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        end
+    end
+
+    return train_data
+end
+
+-- Determines which locomotive in a train is the main one
+-- @tparam LuaTrain train
+-- @return LuaEntity the main locomotive
+function get_main_locomotive(train)
```
Comment (Afforess) [inline stdlib/trains/trains.lua]: Should this be exposed as part of the API? Train.get_main_locomotive?

File: stdlib/trains/trains.lua [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+-- @tparam LuaTrain train
+-- @return int
+function Trains.get_train_id(train)
+  local loco = get_main_locomotive(train)
+  return loco and loco.unit_number
+end
+
+--- Event fired when some change has happened to a locomotive
+-- @return void
+function Trains.on_locomotive_changed()
```
Comment (Afforess) [inline stdlib/trains/trains.lua]: The name of this function should be prefixed with an underscore to indicate it is an internal (protected) API. E.g: `Trains._on_locomotive_changed()`

File: stdlib/trains/trains.lua [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+            old_id = renaming.old_id,
+            new_id = renaming.new_id
+        }
+        Event.dispatch(Trains.on_train_renamed, event_data)
+    end
+end
+--- Event fired when a new locomotive has been created
+-- @tparam LuaEntity new_locomotive The new entity
+-- @return void
+function Trains.on_locomotive_created(new_locomotive)
```
Comment (Afforess) [inline stdlib/trains/trains.lua]: Same as https://github.com/Afforess/Factorio-Stdlib/pull/46/files#r78116282

File: stdlib/trains/trains.lua [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+-- <p>
+--  <table>
+--      <tr><td>train (LuaTrain)</td><td>The LuaTrain instance</td></tr>
+--      <tr><td>id (int)</td><td>The id of the train</td></tr>
+--  </table>
+-- </p>
+function Trains.find_filtered(criteria)
+    -- Get locomotives by filter
+    surface_index = criteria.surface_name or 1
+    locomotives = Surface.find_all_entities({
```
Comment (Afforess) [inline stdlib/trains/trains.lua]: I'm a bit confused here. The only criteria you've given to `Surface.find_all_entities` is the surface index. Won't this return *all* entities, not just trains? You might want to tweak your tests to add a non-train entity and make sure this is working as you expect it to...

I think you need to qualify the type of entity.

File: stdlib/trains/trains.lua [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+-- @return A list of train details tables, if any are found matching the criteria. Otherwise the empty list.
+-- <p>
+--  <table>
+--      <tr><td>train (LuaTrain)</td><td>The LuaTrain instance</td></tr>
+--      <tr><td>id (int)</td><td>The id of the train</td></tr>
+--  </table>
+-- </p>
+function Trains.find_filtered(criteria)
+    -- Get locomotives by filter
+    surface_index = criteria.surface_name or 1
```
Comment (Afforess) [inline stdlib/trains/trains.lua]: Shouldn't these inner variables be prefixed with `local`?

File: stdlib/trains/trains.lua [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+-- @module Trains
+
+-- Event registration is performed at the bottom of this file,
+-- once all other functions have been defined
+
+require 'stdlib/event/event'
+require 'stdlib/surface'
+require 'stdlib/table'
+
+local registry_loaded = false
```
Comment (Afforess) [inline stdlib/trains/trains.lua]: Dead variable, this should be removed.

File: stdlib/trains/trains.lua [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+
+local Trains = {
```
Comment (Afforess) [inline stdlib/trains/trains.lua]: This should *not* be local. Trains is part of the global namespace.

**Comment by Afforess** [comment]
Mostly a bunch of style comments, with one actual implementation question. Looks pretty good.

Feel free to disagree vigorously with any of my suggestions. I'm just trying to spot check for errors and I don't always get everything right.

**Comment by AndyBursh** [comment]
Those are all fair points, especially given my lack of familiarity with Lua and its idioms. The only one I would question is making `Trains` non-local. If you make it non-local then return it at the bottom of the module, what effect does that have on this statement; `Trains = require('stdlib/trains/trains')`?

The missing `find_filtered` issue is absolutely spot on. I got so caught up in learning everything else, I missed that bit! I'll update my repository, squash the changes in the head of this branch and let you know so you can review it again.

**Comment by Afforess** [comment]
>>The only one I would question is making Trains non-local. If you make it non-local then return it at the bottom of the module, what effect does that have on this statement; Trains = require('stdlib/trains/trains')?

That statement still works, but it would re-assign the global Trains to itself. The returns at the end of each module are mostly for style reasons. Basically, the lua module syntax is more verbose and easier to make a mistake, ex:

```
Position = require 'stdlib/area/position'
Area = require 'stdlib/area/area'
Event = require 'stdlib/event/event'
```

vs the current implementation:

```
require 'stdlib/area/position'
require 'stdlib/area/area'
require 'stdlib/event/event'
```

Its a bit unusual and slightly outside the what Lua recommends for modules, but most factorio modders aren't professionals and the simpler the syntax, the more likely the stdlib will be useful.

**Comment by AndyBursh** [comment]
That all makes sense and is fair enough. I've made the changes you suggested. I wasn't sure about adding `get_main_locomotive` to the `Trains` namespace since it's such a utilitarian function, but I figured there's no harm in exposing it.

Whilst I was at it , I've also made `Trains.find_filtered` support all the criteria that the `Surface` module supports. It will default to finding `diesel-locomotive` entities, but it can be overridden by supplying the `name` key should other mods\the game introduce new types of train engines in the future.

**Comment by AndyBursh** [comment]
I've spotted some typos in the names of events. I will update this PR with corrections to those later today.

File: stdlib/trains/trains.lua [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+require 'stdlib/event/event'
+require 'stdlib/surface'
+require 'stdlib/table'
+
+Trains = {
+    _registry = {}
+}
+
+--- Event id for the custom event fired
+-- when a train has been renamed
```
Comment (Afforess) [inline stdlib/trains/trains.lua]: Can you clarify this statement a little bit? It's fairly important documentation, as without a clear understanding, the event won't be used, and its unclear what exactly 'renaming' a train means. If 'renaming' is the correct terminology, please define it in with regards to factorio.

**Comment by Afforess** [comment]
>I haven't really included any examples of how modders would use this module in conjunction with other modules (eg the Entity module) because I didn't really know where to put them. If you can provide guidance on that, I'll add them.

Examples can go in the examples/ top level directory, by the way.

I made one last comment on a matter of documentation. When that is clarified, I'll merge this in!

**Comment by AndyBursh** [comment]
Sure thing. I'll update this PR later today. Should I raise another issue discuss additions to the module, or would you rather do it all in one PR?

**Comment by Afforess** [comment]
>Sure thing. I'll update this PR later today. Should I raise another issue discuss additions to the module, or would you rather do it all in one PR?

Let's make it a separate issue/PR. It's easier to reason about smaller changes than one big one.

**Comment by AndyBursh** [comment]
Ok. I'll raise the issue soon.

**Pull request merged**